lib/std/http/Client.zig

```diff
@@ -1309,19 +1334,49 @@ pub const basic_authorization = struct {
 
 pub const ConnectTcpError = Allocator.Error || error{ ConnectionRefused, NetworkUnreachable, ConnectionTimedOut, ConnectionResetByPeer, TemporaryNameServerFailure, NameServerFailure, UnknownHostName, HostLacksNetworkAddresses, UnexpectedConnectFailure, TlsInitializationFailed };
 
-/// Connect to `host:port` using the specified protocol. This will reuse a connection if one is already open.
-///
-/// This function is threadsafe.
-pub fn connectTcp(client: *Client, host: []const u8, port: u16, protocol: Connection.Protocol) ConnectTcpError!*Connection {
-    if (client.connection_pool.findConnection(.{
-        .host = host,
-        .port = port,
-        .protocol = protocol,
-    })) |node| return node;
+fn createTlsClient(client: *Client, stream: net.Stream, host: []const u8) error{ OutOfMemory, TlsInitializationFailed }!*std.crypto.tls.Client {
+    if (disable_tls) {
+        unreachable;
+    }
 
-    if (disable_tls and protocol == .tls)
-        return error.TlsInitializationFailed;
+    const tls_client = try client.allocator.create(std.crypto.tls.Client);
+    errdefer client.allocator.destroy(tls_client);
 
+    const ssl_key_log_file: ?std.fs.File = if (std.options.http_enable_ssl_key_log_file) ssl_key_log_file: {
+        const ssl_key_log_path = std.process.getEnvVarOwned(client.allocator, "SSLKEYLOGFILE") catch |err| switch (err) {
+            error.EnvironmentVariableNotFound, error.InvalidWtf8 => break :ssl_key_log_file null,
+            error.OutOfMemory => return error.OutOfMemory,
+        };
+        defer client.allocator.free(ssl_key_log_path);
+        break :ssl_key_log_file std.fs.cwd().createFile(ssl_key_log_path, .{
+            .truncate = false,
+            .mode = switch (builtin.os.tag) {
+                .windows, .wasi => 0,
+                else => 0o600,
+            },
+        }) catch null;
+    } else null;
+    errdefer if (ssl_key_log_file) |key_log_file| key_log_file.close();
+
+    tls_client.* = std.crypto.tls.Client.init(stream, .{
+        .host = .{ .explicit = host },
+        .ca = .{ .bundle = client.ca_bundle },
+        .ssl_key_log_file = ssl_key_log_file,
+    }) catch return error.TlsInitializationFailed;
+    // This is appropriate for HTTPS because the HTTP headers contain
+    // the content length which is used to detect truncation attacks.
+    tls_client.allow_truncation_attacks = true;
+
+    return tls_client;
+}
+
+/// Connect to `host:port` using tcp, return the connection pool node.
+/// The connection's host and port is not initialised.
+///
+/// The result is not fully initialised. To destory it, you must close the `stream` and destory the pointer.
+///
+/// This function is threadsafe.
```

Comment:
why this function thread-safe?

Reply:
This function is originally thread-safe and most of the old code is moved to connectHttpTunnel. I carefully construct the new connectSocks5Tunnel, it follows the original routine and atomizes modifications on the shared state (`*Client`).

Reply:
Oops, I misunderstand your question. `createTcpConnect` does not modify any shared state, so it's thread-safe.